enhancements/deprecated-bundles.md

```diff
@@ -0,0 +1,126 @@
+last-updated: 2020-07-15
+status: implementable
+---
+
+# Deprecating Bundles
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+
+## Summary
+
+OPM will support marking versions as deprecated and truncating the update graph of a given package. Deprecated versions will not be installable. If a deprecated version is installed prior to upgrade appropriate alerts will be fired on cluster. 
+
+## Motivation
+
+The default set of Operators provided on OCP clusters will soon (4.6+) be delivered through indexes and each OCP minor version will have a unique set of index images. Today, packages in one index will be duplicated to the next minor version index when it is available. This means operator owners releasing through the redhat, certified, or community pipelines will now need a way to specify which versions of their operator are incompatible or unsupported in the next minor version index.
+
+### Terminology
+
+#### Deprecated bundle
+
+A deprecated bundle is an upgradable but not installable bundle in the update graph of an operator.
+
+#### Terminated bundle
+
+A terminated bundle is neither upgradable nor installable.
+
+### Goals
+
+- A deprecated bundle in the update graph and any of its replacements (terminated bundles) will not be installed by OLM
+- Operator Owners don’t need to release a new version in order to specify which bundle is deprecated
+- Index maintainers have a way to un-do deprecation if needed
+- The deprecation is specified after promotion (not during) and before publishing of the index image
+- Alerts are fired on cluster when a deprecated version is installed
+
+### Non-Goals
+
+- OLM supports upgrading from a terminated bundle
+- Alerts are fired if terminated bundles exist on cluster
+- OLM prevents cluster upgrades based on whether versions are terminated in the upgraded cluster’s index
+
+## Proposal
+
+We propose to add a command to `opm index` called `deprecate` which takes in a set of bundle images, an index, and all the same arguments as `opm index add`. The command modifies the index by removing all versions that come before the specified bundle image in the update graph (that is, until the tail of each channel). If this operation removes the head of a channel, then the channel is removed. The removal of bundles from the update graph will not modify the update graph after the deprecated version (from the deprecated channel to the head of the channel).
+
+Versions that were removed from the index will not receive updates if subscriptions still exist for them on cluster. Ideally, appropriate communication from the operator maintainer will mean that cluster admins will upgrade to the latest supported version prior to cluster upgrade. In the event that the cluster is upgraded and `Subscriptions` to terminated bundles exist on cluster, the cluster admin can manually trigger an upgrade by either:
```

Comment:
In my experience, the upgrade process seems to happen in the reverse. Cluster admins first upgrade the OCP version and then sometimes remember to update the operator version/channel to match. How would an operator maintainer communicate with a cluster admin that they need to update an operator before updating a cluster? If the cluster is updated first, the operator can detect the newer OCP version and fire an alert to let the admin know they should upgrade before the older operator version gets marked as deprecated. You can't do the reverse - an alert that says, "If you choose to upgrade to OCP 4.x you'll need to update this operator first" is just noise as you have no idea if or when the cluster admin would choose to do that upgrade.

Reply:
I completely agree. The communication I am expecting from operator maintainer to cluster admin is, so far, just in the form of documentation for supporting manual subscriptions. I guess the way this is phrased makes it seem like the latter workflow is less likely so I can re-phrase that.

Reply:
Offline we discussed:

1. An RFE, by which a user could have OLM do the alerting based on a channel
2. An operator marked for deprecation at a particular version can start to fire alerts. This is has a drawback where if a user does not upgrade to that particular version, then you don't get the alerts.